// oracle-handler/index.js
if (process.platform == 'linux') {
	var oracle = require('oracle');
}
else if (process.platform == 'darwin') {
	var oracle = require('oracle_mac');
}

var connectData = {
    hostname: "localhost",
    port: 1525,
    database: "crs", // System ID (SID)
    user: "tfung",
    password: "pass2014"
};

function oracleQuery(statement) {
	oracle.connect(connectData, function(err, connection) {
	if (err) { console.log('Error connecting to db:', err); return; }

		connection.execute(statement, [], function(err, results) {
			if (err) { console.log('Error executing query:', err); return; }

			console.log(results);
			connection.close(); // call only when query is finished executing
		});
	});
}

function oracleRetrieval(statement, cb) {
	var results = [];

	oracle.connect(connectData, function(err, connection) {
		connection.setPrefetchRowCount(50);

		var reader = connection.reader(statement, []);

		function doRead(cb) {
			reader.nextRow(function(err, row) {
				if (err) return cb(err, null);
				if (row) {
					console.log("got " + JSON.stringify(row));
					results.push(row);

					// recurse to read next record
					return doRead(cb);
				} else {
					// we are done
					console.log("all records processed");
					return cb(null, results);
				}
			});
		}

		doRead(cb);
	});	
}

exports.oracleQuery = oracleQuery;
exports.oracleRetrieval = oracleRetrieval;
