// oracle-handler/index.js
// should probably be moved into helper folder to consolidate 
// custom functions

if (process.platform == 'linux') {
	var oracle = require('oracle');
}
else if (process.platform == 'darwin') {
	var oracle = require('oracle_mac');
}

var connectData = {
    hostname: "localhost",
    port: 1525,
    database: "crs", // System ID (SID)
    user: "tfung",
    password: "pass2014"
};

var oracleDataTypes = [
	'SYSTIMESTAMP'
]

// unsure of how to return success and failure in javascript
function oracleQuery(statement) {
	oracle.connect(connectData, function(err, connection) {
		if (err) { 
			console.log('Error connecting to db:', err); 
			console.log(statement);

			return 0; 
		}

		connection.execute(statement, [], function(err, results) {
			if (err) { 
				console.log('Error executing query:', err); 
				console.log(statement);
				connection.close();

				return 0; 
			}

			console.log(results);
			connection.close(); // call only when query is finished executing
		});
	});

	return 1;
}

function oracleInsert(table, jsonData) {
	var tableString;
	var valueString;

	for (var key in jsonData) {
		var value = oracleDataTypes.indexOf(jsonData[key]) < 0 ? '\''+jsonData[key]+'\'' : jsonData[key];

		if (tableString != undefined) {
			var temp = oracleDataTypes.indexOf(jsonData[key]) < 0 ? '\''+jsonData[key]+'\'' : jsonData[key];

			tableString = tableString + ',' + key;
			valueString = valueString + ',' + value;
		}
		else {
			tableString = key;
			valueString = value;
		}
	}

	var insertStatement =
		'INSERT INTO '+table+' ('+tableString+') '+
		'VALUES ('+valueString+')';

	return(oracleQuery(insertStatement));
}

function oracleRetrieval(statement, cb) {
	var results = [];

	oracle.connect(connectData, function(err, connection) {
		connection.setPrefetchRowCount(50);

		var reader = connection.reader(statement, []);

		function doRead(cb) {
			reader.nextRow(function(err, row) {
				if (err) return cb(err, null);
				if (row) {
					var result = JSON.stringify(row);
					console.log("got " + result);
					results.push(result);

					// recurse to read next record
					return doRead(cb);
				} else {
					// we are done
					console.log("all records processed");
					return cb(null, results);
				}
			});
		}

		doRead(cb);
	});	
}

exports.oracleQuery = oracleQuery;
exports.oracleInsert = oracleInsert;
exports.oracleRetrieval = oracleRetrieval;
